File: bestia-client/src/Bnet/Message/Inventory/UseItemCMSG.cs
using Bnet;

namespace BestiaBehemothClient.Bnet.Message.Inventory
{
  public partial class UseItemCMSG : ICMSG
  {
    public ulong ItemId { get; set; }

    /// <summary>What the client gathered first, for an item whose script needs it. Null for a plain use.</summary>
    public Message.ScriptArgs Args { get; set; }

    public override Envelope ToEnvelope()
    {
      var useItem = new global::Bnet.UseItemCMSG
      {
        ItemId = ItemId,
        Args = Args?.ToProto()
      };

      return new Envelope
      {
        UseItem = useItem
      };
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/item/UseItemCMSG.kt
package net.bestia.zone.item

import net.bestia.bnet.proto.UseItemCMSGProto
import net.bestia.zone.message.CMSG
import net.bestia.zone.script.ScriptArgs

data class UseItemCMSG(
  override val playerId: Long,
  val itemId: Long,

  /** Whatever the client gathered first, for an item whose script asks for more than "use it". */
  val args: ScriptArgs
) : CMSG {

  companion object {
    fun fromBnet(playerId: Long, bnet: UseItemCMSGProto.UseItemCMSG): UseItemCMSG {
      return UseItemCMSG(
        playerId = playerId,
        itemId = bnet.itemId,
        args = ScriptArgs.fromBnet(bnet.args)
      )
    }
  }
}
